rtrlib/rtr_mgr.c

```diff
@@ -91,135 +120,171 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
+		    (current_group != my_group) &&
+		    (current_group->preference > my_group->preference)) {
+			for (unsigned int j = 0;
+				j < current_group->sockets_len; j++) {
 				pthread_mutex_unlock(&config->mutex);
-				rtr_stop(cg.sockets[j]);
+				rtr_stop(current_group->sockets[j]);
 				pthread_mutex_lock(&config->mutex);
 			}
-			set_status(config, &cg, RTR_MGR_CLOSED, sock);
+			set_status(config, current_group, RTR_MGR_CLOSED, sock);
 		}
+	node = node->next;
 	}
+	pthread_mutex_unlock(&config->mutex);
 }
 
-static struct rtr_mgr_group
-*get_best_inactive_rtr_mgr_group(struct rtr_mgr_config *config,
-				 unsigned int my_group_idx)
+static struct rtr_mgr_group *get_best_inactive_rtr_mgr_group(
+						struct rtr_mgr_config *config,
+						struct rtr_mgr_group *my_group)
 {
-	/* groups are sorted by preference */
-	for (int i = 0; i < config->len; i++) {
-		if ((i != my_group_idx) &&
-		    (config->groups[i].status == RTR_MGR_CLOSED))
-			return &config->groups[i];
+	struct rtr_mgr_group_node *group_node;
+	struct rtr_mgr_group *group;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+	group_node = node->data;
+	group = group_node->group;
+		if ((group != my_group) && (group->status == RTR_MGR_CLOSED)) {
+			pthread_mutex_unlock(&config->mutex);
+			return group;
+		}
+	node = node->next;
 	}
+		pthread_mutex_unlock(&config->mutex);
 	return NULL;
 }
 
 static bool is_some_rtr_mgr_group_established(struct rtr_mgr_config *config)
 {
-	for (int i = 0; i < config->len; i++) {
-		if (config->groups[i].status == RTR_MGR_ESTABLISHED)
+	struct rtr_mgr_group_node *group_node;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		group_node = node->data;
+		if (group_node->group->status == RTR_MGR_ESTABLISHED) {
+			pthread_mutex_unlock(&config->mutex);
 			return true;
+		}
+		node = node->next;
 	}
+	pthread_mutex_unlock(&config->mutex);
 	return false;
 }
 
 static inline void _rtr_mgr_cb_state_shutdown(const struct rtr_socket *sock,
 					      struct rtr_mgr_config *config,
-					      unsigned int ind)
+					      struct rtr_mgr_group *group)
 {
 	bool all_down = true;
 
-	for (unsigned int i = 0; i < config->groups[ind].sockets_len; i++) {
-		if (config->groups[ind].sockets[i]->state != RTR_SHUTDOWN) {
+	for (unsigned int i = 0; i < group->sockets_len; i++) {
+		if (group->sockets[i]->state != RTR_SHUTDOWN) {
 			all_down = false;
 			break;
 		}
 	}
 	if (all_down)
-		set_status(config, &config->groups[ind],
+		set_status(config, group,
 			   RTR_MGR_CLOSED, sock);
 	else
-		set_status(config, &config->groups[ind],
-			   config->groups[ind].status, sock);
+		set_status(config, group,
+			   group->status, sock);
 }
 
 static inline void _rtr_mgr_cb_state_established(const struct rtr_socket *sock,
 						 struct rtr_mgr_config *config,
-						 unsigned int ind)
+						 struct rtr_mgr_group *my_group)
```

Comment:
inconsistent parameter naming, above its just `group` which is fine - `my_group` looks a bit like code from a beginners tutorial 😉